### PR TITLE
Fix the printf implementation for NULL pointers

### PR DESCRIPTION
**Pull request by stefanberger**
## Description of the changes 

Thie PR fixes the printf implementation for NULL pointers.

---
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/165)

### CODE REVIEW

**Review by dimakuv — APPROVED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/165)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @boryspoplawski)

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/165)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/165#-MmXruUQ2mAQCvD5xUck):*
<details><summary><i>Previously, stefanberger (Stefan Berger) wrote…</i></summary><blockquote>

I had NULL pointer crashes in proc_common.c and proc_cpuinfo.c testcases on ppc64 in the fopen() call (weird).  Even though it works now on ppc64, I think I may have to send another PR to pass va_list *ap rather than va_list ap. We also had to do this in shim_parser.c all over the place...
</blockquote></details>

Hm, but `vsnprintf` has kind of unchangeable interface (and uses `va_list ap`)...

**Review by boryspoplawski — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/165)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/gramineproject/gramine/165#-MmXruUQ2mAQCvD5xUck):*
<details><summary><i>Previously, stefanberger (Stefan Berger) wrote…</i></summary><blockquote>

So this patch here is correct by itself, so I think you should be able to merge it.

I think the troubles are starting when the `ap` parameter is passed down into a function that starts accessing `ap` and as part of that probably manipulates `ap` itself and then again is passing `ap` into yet another function that again accesses and manipulated `ap`. So in this case it's better to have this be a pointer, however, after investigation of the new code, this isn't the case here. I also modified the new printf implementation to use `*ap` and still have the crashes if I don't apply this patch here. It's so weird that this happens somewhere in the code path that fopen() is executing. For not let's forget about this `*ap` thing.
</blockquote></details>

Yeah, we merge it for sure (just waiting for one more approval, as we need 3 to merge it that fast).

As far as I remember the issues we had before your explanation is correct and also as you noted it's not the case here.

**Review by mkow — DISMISSED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/165)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by boryspoplawski — APPROVED** [review]
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/165)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL)

**Review by dimakuv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/165)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**
